include/utils/actions/dfs_sort.h

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2022 Ultimaker B.V.
+// CuraEngine is released under the terms of the AGPLv3 or higher
+
+#ifndef CURAENGINE_DFS_SORT_H
+#define CURAENGINE_DFS_SORT_H
+
+#include <range/v3/algorithm/contains.hpp>
+#include <range/v3/view/subrange.hpp>
+
+#include "utils/concepts/graph.h"
+
+namespace cura::actions
+{
+/* # dfs_sort
+ *
+ * Walks through a directed graph using depth-first-search and adds the nodes it visited to the visited set
+ * \param node the current visited node, should be convertable or have the same type as the nodes in dag and visited (currently not constrained by a `require`)
+ * \param dag Directed Acyclic Graph as defined by \concept isGraph
+ * \param visited nodes as defined by concept \isSet; _note: visited will not be ordered if it is of type `unordered_###`
+ */
+constexpr void dfs_sort(const auto& node, const isGraph auto& dag, isSet auto& visited)
+{
+    if (ranges::contains(visited, node))
+    {
+        return;
+    }
+    visited.push_back(node);
+    const auto& [children_begin, children_end] = dag.equal_range(node);
+    auto children = ranges::make_subrange(children_begin, children_end);
+    for (const auto& [_, child] : children)
+    {
+        dfs_sort(child, dag, visited);
+    }
+}
```

Comment:
As we are working with a directed a cyclic graph we know that there cannot be any cycles. As such the visited check is redundant.

The order which we try to calculated is called the [topological sort](https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search) of the nodes

Also note that the algorithm below doesn't work on arbitrary DAG's as stated in the documentation of the code; only because w know that we are working with rooted trees rather then arbitrary DAG's and we start traversing each tree from the known root the algorithm works. (we are working with tree's which is a subset of a dag; $\text{tree} \subset \text{trees} \subset \text{dag} \subset \text{graph}$).
```suggestion
/* # topolocial_sort
 *
 * Walks through a directed graph using depth-first-search and adds the nodes in that order to the the order list argument
 * \param node the current visited node, should be convertable or have the same type as the nodes in dag and visited (currently not constrained by a `require`)
 * \param dag Directed Acyclic Graph as defined by \concept isGraph
 * \param order nodes as defined by concept \isList; _note: nodes are ordered according to the their topological sorting order`
 */
constexpr void topological_sort(const auto& node, const isGraph auto& dag, isList auto& order)
{
    order.push_back(node);
    const auto& [children_begin, children_end] = dag.equal_range(node);
    auto children = ranges::make_subrange(children_begin, children_end);
    for (const auto& [_, child] : children)
    {
        topological_sort(child, dag, order);
    }
}
```

Reply:
Please note that the above code suggestion is just to convey my intention; I haven't really tested it and it will ofcourse not work as the function calls calling `dfs_sort` should also change to `topological_sort`